nn.py

```diff
@@ -0,0 +1,202 @@
+                group_end_indices = np.cumsum(group_sizes) - 1
+                group_indices.append(group_end_indices)
+
+                permutation = np.concatenate(permutation)
+                permutations.append(permutation)
+
+        return permutations, group_indices
+
+    @staticmethod
+    def _aggregate(sales, permutations, group_indices):
+        """Aggregates sales to compute all levels of the time series hierarchy.
+
+        Args:
+            sales         = [torch.Tensor] Tensor of unit or dollar sales
+            permutations  = [[np.ndarray]*12] sales permutation for each group
+            group_indices = [[np.ndarray]*12] end indices for each group
+
+        Returns [torch.Tensor]:
+            Tensor of aggregate sales given the permutations and group indices.
+        """
+        aggregates = []
+        for permutation, group_end_indices in zip(permutations, group_indices):
+            # permute input to get consecutive groups
+            permutation = sales[permutation]
+
+            # compute cumulative sum of sales along each group
+            sums1 = permutation.cumsum(0)[group_end_indices]
+            sums2 = torch.cat((torch.zeros_like(sales[:1]), sums1[:-1]))
+
+            # add aggregate sum of sales to list
+            aggregates.append(sums1 - sums2)
+
+        return torch.cat(aggregates)
+
+    def forward(self, input, target):
+        """Computes the WRMSSE loss.
+
+        Args:
+            input  = [torch.Tensor] projected unit sales with shape (N, h)
+            target = [torch.Tensor] actual unit sales with shape (N, h)
+
+        Returns [torch.Tensor]:
+            Tensor with a single value for the loss.
+        """
+        # aggregate the data to all levels of the time series hierarchy
+        projected_sales = self._aggregate(
+            input, self._permutations, self._group_indices
+        )
+        target = target.to(self._device)
+        actual_sales = self._aggregate(
+            target, self._permutations, self._group_indices
+        )
+
+        # compute WRMSSE loss
+        squared_errors = (actual_sales - projected_sales)**2
+        MSE = torch.sum(squared_errors, dim=1) / input.shape[1]
+        RMSSE = torch.sqrt(MSE / self._scales)
+        loss = torch.sum(self._weights * RMSSE)
+
+        return loss
+
+
+if __name__ == '__main__':
+    from datetime import datetime
+
+    path = ('D:\\Users\\Niels-laptop\\Documents\\2019-2020\\Machine Learning in'
+            ' Practice\\Competition 2\\project\\')
+    calendar = pd.read_csv(path + 'calendar.csv')
+    prices = pd.read_csv(path + 'sell_prices.csv')
+    sales = pd.read_csv(path + 'sales_train_validation.csv')
+
+    device = torch.device('cpu')
+    time = datetime.now()
+    criterion = WRMSSE(device, calendar, prices, sales)
+    print('Time to initialize loss: ', datetime.now() - time)
+
+    input = torch.rand(30490, 5, device=device)
+    target = torch.rand(30490, 5)
```

Comment:
Does `5` represent the length of each input timeseries here (i.e., 5 days)?

Reply:
Does the new commit answer your question?

Reply:
Yes, I was actually a bit confused about the `input` variable name as it is actually the supposed output of the model (predicted sales/forecast).

Reply:
I usually take over the names of the parameters that are used in the same method in the super class. I do agree that it's confusing, but it does fit better within the PyTorch framework.

I will merge this PR.